### PR TITLE
Bump the `Cake.Core` package to 4.0.0

### LINKED ISSUE #123: Outdated version 1.0.0 of referenced `Cake.Core` package
The [Cake.Core](https://github.com/cake-build/cake) version [4.0.0](https://github.com/cake-build/cake/releases/tag/v4.0.0) was released on Nov 19, 2023❕

My local and CircleCi build log shows the following messages:
```
The assembly 'Cake.Coveralls, Version=1.1.0.0, Culture=neutral, PublicKeyToken=null' is referencing an older version of Cake.Core (1.0.0). 
For best compatibility it should target Cake.Core version 4.0.0.
```

## Expected Behavior
[Cake.Core](https://github.com/cake-build/cake) version [4.0.0](https://github.com/cake-build/cake/releases/tag/v4.0.0)

## Current Behavior
[Cake.Core](https://github.com/cake-build/cake) reference is [1.0.0](https://github.com/search?q=repo%3Acake-contrib%2FCake.Coveralls+%22Include%3D%5C%22Cake.Core%5C%22+Version%3D%5C%221.0.0%5C%22%22&type=code)

image

## Possible Solution
Bumping version to the latest is required.

## Steps to Reproduce (for bugs)
1. Search for references [1.0.0](https://github.com/search?q=repo%3Acake-contrib%2FCake.Coveralls+%22Include%3D%5C%22Cake.Core%5C%22+Version%3D%5C%221.0.0%5C%22%22&type=code)

## Context
I see the warnings in CircleCI logs when using [build.cake](https://github.com/ThreeMammals/Ocelot/blob/develop/build.cake).
See latest builds: [CircleCI | ThreeMammals/Ocelot](https://app.circleci.com/pipelines/github/ThreeMammals/Ocelot)
Or look into latest PR builds for [develop](https://github.com/ThreeMammals/Ocelot/commits/develop) branch

## Your Environment
* Addin version used: Cake.Coveralls&version=1.1.0
* Cake Version used: 1.1.0 and 4.0.0
* Operating System: Linux, Windows

### PR DESCRIPTION
**Pull request by raman-m**
## Fixes #123 
- #123 

This PR upgrades the [Cake.Core](https://github.com/cake-build/cake) version to [4.0.0](https://github.com/cake-build/cake/releases/tag/v4.0.0) which was released on Nov 19, 2023.
With minor refactoring, with checking and fixing errors/warnings/messages of StyleCop and IDE code analyzers.

## Description
- Bumped `Cake.Core` to [4.0.0](https://github.com/cake-build/cake/releases/tag/v4.0.0) and other packages to the latest versions, see commit https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/0407e042f5ed555d1bc3190d1c88aff9941921bf
- Removed unsupported `Cake.Addin.Analyzer`, the latest version [0.1.3](https://www.nuget.org/packages/Cake.Addin.Analyzer/0.1.3) was [released on Mar 3, 2021](https://github.com/WormieCorp/Cake.Addin.Analyzer/releases), see commit https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/acaa3f7293cff11b14d6fcbf6ae128188625566e
- Fixed the first StyleCop warnings, see commits https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/c44057ababb8122c00ddc0f6219a7d5262c70ee7 and https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/c88a432802864915c511f0636833909582eff5f9
- Added `.editorconfig` to the solution because it is best practice for code analysis, see commit https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/42aa7f719c44b5c30ae4c3810fff4ab9e83065c3
- Fixed Cake Contrib Guidelines analysis error `CCG0008` related to usage of recommended tags, see commit https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/634170be6bfa2a01960673c0ce8750589a0d52c8
- Setup code analysis and StyleCop having updated solution & project files, see commit https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/bd71595b37b86c6fa0d8b60913df9a587e995166
- Fixed StyleCop warnings: `SA1210` (https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/2c473dbe1f43924504eddf82dfceea58e9d7848d), `SA1518` (https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/fb0c16a5d2c84fbb4511a0c9136e7d3c587bf315), `SA1600` (https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/3ed30e3cd01fe9a73a466b5eeada1e6350e205c9) and `SA0001` (https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/f3520549e76d199cdc8edf1dcf84de79299694bc)
- Fixed CSharp analysis messages for testing project: `CS1591` (https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/206cd630282ce5d0bb1a32d2f58609dc031cc3f2)
- Followed IDE Code Analysis suggestion: `CA1510` (https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/1ffc8d3176aea34fdfc5bfa381dcff7291dfa334)
- Followed IDE Code Analysis suggestions: `IDE0017` (https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/d559dedb372d8d410f68c2c91030daa2b41b0314) and `IDE0066` (https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/778f8662a9026e54108a123f3e29e10f0980d049)
- Disabled 2 features which are available in .NET 8 (C# 12) only by IDE Code Analysis suggestions: `IDE0290` (https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/c0803097de328f1650daf1d139fda1163a599016) and `IDE0300` (https://github.com/cake-contrib/Cake.Coveralls/pull/124/commits/9efd4822ef56e053e784c50146335b878c244775)

## Related Issue
- Fixes #123 

## Motivation and Context
This change is required because of #123 
After major release of referenced packages, `Cake.Core` one in our case, the packs versions must be updated.

## How Has This Been Tested?

To be provided...

## Screenshots (if appropriate):
image

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes. **Note** by @raman-m: No actual changes means no new tests.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by raman-m** [comment]
@gep13 Hey Gary!

I’d appreciate it if you could find some time to review, merge into the develop branch, and then release the next version, please!

**Comment by raman-m** [comment]
@devlead Hello Mattias!
Do you know who maintains the repository currently?

### Workflow status
> 1 workflow awaiting approval
This workflow requires approval from a maintainer.

**Comment by raman-m** [comment]
@gep13  Gary
@devlead Mattias
Another notification has been issued for a new release due to Cake 4.0. I am confident that my pull request is of sufficient quality to be merged and released.

**Comment by raman-m** [comment]
@gep13 @nmbro @Jericho @daveaglick @derhally @augustoproiete @Sharparam @devlead @larzw @RLittlesII @jokay 
Hello everyone, I apologize for directly tagging all of you!
Could anyone inform me who currently owns this repository? Who is responsible for reviewing PRs and managing releases?
Thank you kindly for directing me to the appropriate GitHub individuals.

File: Source/Cake.Coveralls.Tests/Cake.Coveralls.Tests.csproj
Comment (nmbro) [inline Source/Cake.Coveralls.Tests/Cake.Coveralls.Tests.csproj:30]: should that be a beta version?

**Review by nmbro — APPROVED** [review]
changes look good at a glance.

I remember @pascalberger being pretty active in the cake-contrib repos

And for the love of everything holy - remember to squash if this gets merged :)

File: Source/Cake.Coveralls.Tests/Cake.Coveralls.Tests.csproj
Comment (raman-m) [reply Source/Cake.Coveralls.Tests/Cake.Coveralls.Tests.csproj:30]: @nmbro  Nicolai,
Thanks for the review!

Yes, It should!
The [NuGet Gallery | StyleCop.Analyzers](https://www.nuget.org/packages/StyleCop.Analyzers/#versions-body-tab) lists the most recent version as [1.2.0-beta.556](https://www.nuget.org/packages/StyleCop.Analyzers/1.2.0-beta.556).

Concerning the repository [DotNetAnalyzers/StyleCopAnalyzers](https://github.com/DotNetAnalyzers/StyleCopAnalyzers), the most recent release is version [1.2.0-beta.556](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/releases/tag/1.2.0-beta.556). It's unclear why the StyleCopAnalyzers team designates releases as Pre-Release (beta). The last non-beta release occurred on [on Apr 29, 2019](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/releases/tag/1.1.118). One reason might be the significant changes in C# language syntax from 2020 to 2024. It's possible that the team is overwhelmed with tasks and lacks the time for comprehensive testing or to gather sufficient feedback from the community, leading to only Pre-Releases being issued.

**Comment by nils-a** [comment]
@raman-m thank you for the contribution. 
I have spoken to @gep13 and I will start bringing this repo up-to-speed in the coming days.

However, rather than doing one big jump to Cake 4, I will do small steps and create intermediate releases for Cake 2.0 and 3.0 between. 
I will do that, since there are people out there stuck on one version or another for different reasons..

If it's ok for you, I'll probably cherry-pick some of your commits along the way..

**Comment by raman-m** [comment]
@nils-a  Hi Nils!
>However, rather than doing one big jump to Cake 4, **I will do small steps and create intermediate releases for Cake 2.0 and 3.0 between**.

Indeed, releasing previous versions of Core is a sensible approach.

> If it's ok for you, I'll probably cherry-pick some of your commits along the way..

That's fine! There's no need to cherry-pick. I'll rebase the feature branch following the official release for versions 2 and 3. However, do let me know if you require any assistance.

P.S. I will resolve conflicts later after the releases.

File: Source/Cake.Coveralls/Cake.Coveralls.csproj
Comment (nils-a) [inline Source/Cake.Coveralls/Cake.Coveralls.csproj:35]: Is there a reason you removed `Cake.Addin.Analyzer`?

**Comment by nils-a** [comment]
@raman-m it is done: I released [Cake.Coveralls 3.0.0](https://github.com/cake-contrib/Cake.Coveralls/releases/tag/3.0.0) which added support for Cake 3.0.0. If you like, you can start resolving conflicts, now. 😬 I hope I have not made too many changes that collide with this PR.

File: Source/Cake.Coveralls/Cake.Coveralls.csproj
Comment (raman-m) [reply Source/Cake.Coveralls/Cake.Coveralls.csproj:35]: The package in question is indeed outdated; the latest version is [0.1.3](https://www.nuget.org/packages/Cake.Addin.Analyzer/0.1.3#versions-body-tab), released [on 03.03.21](https://github.com/WormieCorp/Cake.Addin.Analyzer/releases). I removed it due to its lack of support for .NET 6+, as critical errors were encountered during project solution upgrades.
Consequently, it fails to analyze the Cake infrastructure accurately within the Visual Studio environment.
Additionally, this package is not an official one.

**Comment by raman-m** [comment]
Going to rebase the feature branch and resolve merge conflicts...

File: Source/Cake.Coveralls/Cake.Coveralls.csproj
Comment (nils-a) [reply Source/Cake.Coveralls/Cake.Coveralls.csproj:35]: Ok. I can follow your reasoning.

I'll create an issue "here" that we re-add it, once it has been updated (I agree, it is also one of those repos that has not have much love lately) and I'll also raise an issue "there" about the fact that is no longer working.

**Comment by raman-m** [comment]
@nils-a, if it's alright with you, I can implement the Visual Studio refactoring Quick Action, also known as "[Convert to file-scoped namespace](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-10#file-scoped-namespace-declaration)", across all C# files. [File-scoped namespaces](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/file-scoped-namespaces) have been available since .NET 6 vs [C# 10](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-10).
What are your thoughts?

### P.S. FYI
The feature branch has been rebased onto [cake-contrib:develop](https://github.com/cake-contrib/Cake.Coveralls/tree/develop)❕

**Comment by nils-a** [comment]
While I, personally, like it, I'd rather have @gep13 input on that question. 
Also, if we're going to implement it, I'd rather have it as an individual PR.

So, no, for now let's not do that.

**Review by nils-a — APPROVED** [review]
LGTM!

**Comment by raman-m** [comment]
Should I close conversation threads?

**Comment by nils-a** [comment]
@raman-m your changes have been merged, thanks for your contribution :+1:

I will fix two more issues, and then I will release the v4.0.0.

**Comment by gep13** [comment]
> @nils-a said...
While I, personally, like it, I'd rather have @gep13 input on that question.

I don't have any strong opinion either way.  Personally, I don't really like it, but I am an old .NET Developer, who is stuck in their ways!  Long live `string.Format`

> @nils-a said...
Also, if we're going to implement it, I'd rather have it as an individual PR.

I think that this would be the correct approach.

**Comment by raman-m** [comment]
Thank you for merging it!
I am looking forward to the v4 release so I can enable and test the new version with the build script in my project.

**Pull request merged**
